csharp/src/Apache.Arrow/Ipc/ArrowFileReader.cs

```diff
@@ -13,203 +13,41 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-using FlatBuffers;
-using System;
-using System.Buffers.Binary;
 using System.IO;
-using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace Apache.Arrow.Ipc
 {
+    /// <summary>
+    /// Implements an <see cref="ArrowStreamReader"/> that can read Arrow files.
+    /// </summary>
     public class ArrowFileReader : ArrowStreamReader
     {
+        private ArrowFileReaderImplementation Implementation =>
+            (ArrowFileReaderImplementation)_implementation;
 
-        public bool IsFileValid { get; protected set; }
+        public bool IsFileValid => Implementation.IsFileValid;
 
-        /// <summary>
-        /// When using GetNextRecordBatch this value 
-        /// is to remember what index is next
-        /// </summary>
-        private int _recordBatchIndex;
-
-        /// <summary>
-        /// Notes what byte position where the footer data is in the stream
-        /// </summary>
-        private int _footerStartPostion;
-
-        private ArrowFooter _footer;
-
-        public async Task<int> RecordBatchCountAsync()
+        public ArrowFileReader(Stream stream)
+            : base(new ArrowFileReaderImplementation(stream))
         {
-            if (!HasReadSchema)
-            {
-                await ReadSchemaAsync();
-            }
-
-            return _footer.RecordBatchCount;
-        }
-
-        public ArrowFileReader(Stream stream) : base(stream)
-        {
-            if (!stream.CanSeek)
-            {
-                throw new ArgumentException("Stream must be seekable.", nameof(stream));
-            }
-
-            if (!stream.CanRead)
-            {
-                throw new ArgumentException("Stream must be readable.");
-            }
-
-            IsFileValid = false;
         }
 
         public static ArrowFileReader FromFile(string filename)
         {
-            var stream = new FileStream(filename, FileMode.Open);
+            var stream = new FileStream(filename, FileMode.Open, FileAccess.Read);
             return new ArrowFileReader(stream);
         }
 
-        protected override async Task<Schema> ReadSchemaAsync()
+        public Task<int> RecordBatchCountAsync()
```

Comment:
Consider making these return `ValueTask<T>` instead.

Reply:
Additionally, I am not sure RecordBatchCountAsync is a good design choice. Would it make more sense to expose this is a property, and extract it when the file is opened or the footer is read for the first time and throw InvalidOperationException otherwise? The intent of this method, I believe, was to read the number of record batches stored in the file (as read from the footer), so you could seek to a specific record batch, or read them one at a time.

Reply:
@stephentoub - is there good documentation on when to use which? I can change all these public APIs over to `ValueTask<T>`. But it feels like a breaking change....

@chutchinson - maybe a new JIRA should be logged for that?

Reply:
I've opened https://issues.apache.org/jira/browse/ARROW-4717 to change these public APIs to use `ValueTask<T>` instead.

Reply:
> is there good documentation on when to use which?

The best right now is this:
https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/

In short:
- Start with `Task and `Task<T>`.  They provide the best usability.  If you don't care about getting the best performance, aren't on a hot path, etc., you're done.
- If the method in question returns a T value and has a high likelihood of completing synchronously, strongly consider using `ValueTask<T>`, in order to avoid the `Task<T>` allocation per `T`.  There are a few exceptions to this, namely cases where you or the runtime can cache every possible or likely T to be returned, e.g. if T == bool.
- If the method in question is willing to sign up for increased complexity and has an effective/efficient means for pooling objects, consider using `ValueTask`/`ValueTask<T>`; the implementation can return a `ValueTask`/`ValueTask<T>` that wraps an `IValueTaskSource`/`IValueTaskSource<T>` which it reuses over and over for many calls.
- If the method in question is virtual or abstract, consider whether some implementer may feel strongly about one of the above cases.

As an example, Stream.Read/WriteAsync now have overloads that return `ValueTask<int>` and `ValueTask`, respectively.  `MemoryStream`'s overrides of these always complete synchronously, so those overrides never need to allocate tasks.  `NetworkStream`'s overrides these and uses a pooled `IValueTaskSource` implementation to avoid allocation even for asynchronous completion, but at the cost of non-trivial implementation complexity.